src/parser.rs

```diff
@@ -693,15 +695,82 @@ fn parse_single_quoted_string(input: &str) -> ParseResult<&str> {
 
 fn parse_double_quoted_string(input: &str) -> ParseResult<Vec<WordPart>> {
   // https://pubs.opengroup.org/onlinepubs/009604499/utilities/xcu_chap02.html#tag_02_02_03
-  // Double quotes may have escaped
-  delimited(
-    ch('"'),
+  parse_surrounded_expression(
+    input,
+    '"',
+    "Expected closing double quote.",
     parse_word_parts(ParseWordPartsMode::DoubleQuotes),
-    with_failure_input(
-      input,
-      assert_exists(ch('"'), "Expected closing double quote."),
-    ),
-  )(input)
+  )
+}
+
+fn parse_surrounded_expression<'a, TResult>(
+  input: &'a str,
+  surrounded_char: char,
+  fail_message: &str,
+  parse: impl Fn(&str) -> ParseResult<TResult>,
+) -> ParseResult<'a, TResult> {
+  let start_input = input;
+  let (input, _) = ch(surrounded_char)(input)?;
+  let mut was_escape = false;
+  for (index, c) in input.char_indices() {
+    match c {
+      c if c == surrounded_char && !was_escape => {
+        let inner_input = &input[..index];
+        let inner_input =
+          if surrounded_char == '`' && inner_input.contains("\\`") {
+            Cow::Owned(inner_input.replace("\\`", "`"))
+          } else {
+            Cow::Borrowed(inner_input)
+          };
+        let parts = match parse(&inner_input) {
+          Ok((result_input, parts)) => {
+            if !result_input.is_empty() {
+              return ParseError::fail(
+                input,
+                format!(
+                  "Failed parsing within {}. Unexpected character: {}",
+                  if c == '`' {
+                    "backticks"
+                  } else {
+                    "double quotes"
+                  },
+                  result_input
+                ),
+              );
+            }
+            parts
+          }
+          Err(err) => {
+            return ParseError::fail(
+              input,
+              format!(
+                "Failed parsing within {}. {}",
+                if c == '`' {
+                  "backticks"
+                } else {
+                  "double quotes"
+                },
+                match &err {
+                  ParseError::Backtrace => "Could not determine expression.",
+                  ParseError::Failure(parse_error_failure) =>
+                    parse_error_failure.message.as_str(),
+                }
+              ),
+            );
```

Comment:
These errors are slightly sloppy because it may be parsing an owned value, but this is good enough for now. We can improve this over time.